### ISSUE TITLE
Azimuthal integral of DFT is not differentiable

### ISSUE THREAD

**Issue by yshinya6** [cyq373/SSD-GAN#2]
Hi. Thank you for sharing your code.
I've found that your implementation of the averaged azimuthal integral of DFT is not differentiable.
It uses Numpy, and thus, the gradient updates using this code seem to affect nothing for the entire training.

https://github.com/cyq373/SSD-GAN/blob/main/get_fft_feature.py#L25-L43

I've also found some other problems with your code:

1.  [In line 46 of get_fft_feature.py](https://github.com/cyq373/SSD-GAN/blob/main/get_fft_feature.py#L46), the input images are detached from the computation graphs; this means there is no gradient for updating generators.

2. [In line 54 of get_fft_feature.py](https://github.com/cyq373/SSD-GAN/blob/main/get_fft_feature.py#L54), why do you use `magnitude_spectrum.cpu().numpy()`?. I think this also stops the backward computation.

After reading your code, I wondered why your implementation improves the baseline.
In my guess, the improvement may mainly come from just increasing the model complexity, i.e., it just delays the iteration for reaching the optimal point of GANs.
I'm validating my guess in my environment, and I'll be in touch with you here when the experiments complete.

Many thanks.

**Comment by cyq373** [comment]
Hi, thank you for your interest in our work.
In our paper (please refer to the section 'Analyzing the Effect of the Spectral Classifier'), we propose that the backpropagation process of the adversarial loss does not pass through the spectral classifier C, and C(φ(x)) serves as a spectral modulating factor to the adversarial loss.

**Comment by yshinya6** [comment]
Thank you for replying. I apologize for misreading your paper.
I've checked Eq.(11-12) and convinced that the effects of the spectral modulating factor for G and D.
I'm very curious about why backpropagations from C degrade the performance (is it azimuthal integral matter?).

I appreciate you if you answer me one more question: 
Did you use another code of differentiable azimuthal integral for applying spectral regularization requiring backpropagations in the case of combining SSD-GAN and spectral regularization?

Lastly, I'd like to share my test results of your work.
I compared SNGAN and SSD-GAN on CIFAR-100 by using [torch-mimicry](https://github.com/kwotsin/mimicry/blob/master/torch_mimicry/nets/sngan/sngan_32.py) and your repository. 
I tested three times and report the mean FID scores as follows:

  | CIFAR-100-32
-- | --
SNGAN | 20.55
SSD-GAN | 19.56

On your paper, the FID score of SSD-GAN is reported as 19.28 so that I can confirm your code have certain reproducibility.

Thanks.

**Comment by cyq373** [comment]
Thank you for double-checking the reproducibility of our code.

From the experiments in the paper, we found that when using discrete wavelet transform (DWT) as the downsampling layer to avoid information loss, the performance of SNGAN degrades.  We conjecture that because the DWT downsampling layer remains all the high-frequency information of the input, which contains both details and noises, it is difficult for the model to learn meaningful semantic representation. Our method is a smarter way to utilize high-frequency information of images.

For the differentiable version of azimuthal integral, I notice that there is a torch version of [bincount](https://pytorch.org/docs/stable/generated/torch.bincount.html). Replacing `np.bincount` in the [get_fft_feature.py](https://github.com/cyq373/SSD-GAN/blob/main/get_fft_feature.py) to `torch.bincount` may be helpful.

**Comment by yshinya6** [comment]
Thank you for your kind replies.

> We conjecture that because the DWT downsampling layer remains all the high-frequency information of the input,

That's very interesting. So, your point is inserting the frequency transform layer into models can struggle the training irrespective of the variation of the transform including DFT (which is used in azimuthal integral) and DWT.

I close this issue because I think the discussions are done.
Thank you.
